File: frontend/src/lib/health.ts
// Position health scoring — turns the daily health.json snapshot into a
// caution level for a stock you HOLD.
//
// Deliberately a CAUTION indicator, not an exit signal: docs/EXITS.md tested
// six exit rules and only the RSI>75 trim beat buy-and-hold in both windows —
// stop-loss and SMA-exit rules underperformed simply holding. So this surfaces
// deterioration for a human to judge; it never says "sell".
import type { TickerHealth } from '../types'

export type HealthLevel = 'strong' | 'ok' | 'caution' | 'weak'

export interface HealthReason {
  text: string
  severity: 1 | 2 // 1 = caution, 2 = serious
}

export interface HealthVerdict {
  level: HealthLevel
  label: string
  reasons: HealthReason[]
  positives: string[]
}

const RSI_TRIM = 75          // the one backtest-validated exit hint
const NEAR_SMA_PCT = 2       // "hovering at the line"
const DRAWDOWN_CAUTION = -15 // % off the 1-year peak

/** Grade a held position. `drawdownSinceEntry` is optional (computed client-side
 *  from entry price vs the ticker's recent peak). */
export function assessHealth(h: TickerHealth | undefined): HealthVerdict {
  if (!h) {
    return { level: 'ok', label: 'No data', reasons: [], positives: [] }
  }
  const reasons: HealthReason[] = []
  const positives: string[] = []

  // --- trend: where price sits vs its 200-day line -------------------------
  if (h.vs_sma200_pct !== undefined) {
    if (h.vs_sma200_pct < 0) {
      reasons.push({ text: `Below its 200-day average (${h.vs_sma200_pct.toFixed(1)}%)`, severity: 2 })
    } else if (h.vs_sma200_pct < NEAR_SMA_PCT) {
      reasons.push({ text: `Hovering just above its 200-day average (+${h.vs_sma200_pct.toFixed(1)}%)`, severity: 1 })
    } else {
      positives.push(`Comfortably above its 200-day average (+${h.vs_sma200_pct.toFixed(1)}%)`)
    }
  }
  if (h.vs_sma50_pct !== undefined && h.vs_sma50_pct < 0 && (h.vs_sma200_pct ?? 1) >= 0) {
    // lost the shorter trend while the long one still holds = early warning
    reasons.push({ text: `Lost its 50-day average (${h.vs_sma50_pct.toFixed(1)}%)`, severity: 1 })
  }

  // --- momentum ------------------------------------------------------------
  if (h.macd_bullish === false) {
    reasons.push({ text: 'MACD momentum has turned negative', severity: 1 })
  } else if (h.macd_bullish) {
    positives.push('MACD momentum positive')
  }
  if (h.chg_20d_pct !== undefined && h.chg_20d_pct < -10) {
    reasons.push({ text: `Down ${h.chg_20d_pct.toFixed(1)}% over the last month`, severity: 1 })
  }

  // --- drawdown ------------------------------------------------------------
  if (h.drawdown_pct !== undefined && h.drawdown_pct <= DRAWDOWN_CAUTION) {
    reasons.push({
      text: `${Math.abs(h.drawdown_pct).toFixed(0)}% below its 1-year peak`,
      severity: h.drawdown_pct <= -25 ? 2 : 1,
    })
  }

  // --- overbought: the ONLY exit rule that survived backtesting -------------
  if (h.rsi !== undefined && h.rsi > RSI_TRIM) {
    reasons.push({
      text: `RSI ${h.rsi.toFixed(0)} — overbought; the one exit rule backtests supported (trim, not exit)`,
      severity: 1,
    })
  }

  // --- context -------------------------------------------------------------
  if (h.sector_state === 'lagging') {
    reasons.push({ text: 'Its sector is lagging the market', severity: 1 })
  } else if (h.sector_state === 'leading') {
    positives.push('Sector is leading the market')
  }

  // --- recent alerts (30-day memory, so a missed day still shows) -----------
  for (const w of h.recent_warnings ?? []) {
    const isTrim = w.category === 'rsi_extended'
    reasons.push({
      text: `${isTrim ? 'Take-profit (RSI>75) alert' : 'Bearish signal'} fired on ${w.date}`,
      severity: isTrim ? 1 : 2,
    })
  }

  const serious = reasons.filter((r) => r.severity === 2).length
  const level: HealthLevel =
    serious >= 2 ? 'weak'
      : serious === 1 ? 'caution'
        : reasons.length >= 2 ? 'caution'
          : reasons.length === 1 ? 'ok'
            : 'strong'
  const label =
    level === 'weak' ? 'Weak' : level === 'caution' ? 'Caution' : level === 'ok' ? 'OK' : 'Strong'
  return { level, label, reasons, positives }
}

export const HEALTH_DOT: Record<HealthLevel, string> = {
  strong: '●',
  ok: '●',
  caution: '▲',
  weak: '▼',
}

File: scanner/health.py
"""Per-ticker technical health → frontend/public/data/health.json.

The Portfolio page needs to warn when a HELD stock is *deteriorating*, not just
on the day it crosses a line. Alerts are events (one bar); this is a daily
STATE snapshot for every scanned ticker: where price sits vs its SMAs, momentum,
recent trend, distance below its recent peak, and whether a bearish/trim alert
fired in the last 30 days (so a warning you missed on Tuesday is still visible
on Friday).

Deliberately DISPLAY-ONLY and deliberately not an exit signal: docs/EXITS.md
found that of six tested exit rules only the RSI>75 trim beat buy-and-hold in
both windows — stop-loss and SMA-exit rules underperformed simply holding. The
frontend grades these into caution levels; nothing here changes a verdict.

Rides the daily scan like sectors/forex (failure-isolated) and is byte-stable:
computed purely from the scan's own bars, so a holiday re-run rewrites identical
bytes and stays a no-op commit.
"""

from __future__ import annotations

import datetime as dt
import json
import logging
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from indicators import macd, rsi, sma

logger = logging.getLogger(__name__)

SCANNER_DIR = Path(__file__).resolve().parent
DEFAULT_OUTPUT_DIR = SCANNER_DIR.parent / "frontend" / "public" / "data"
SCHEMA_VERSION = 1
PEAK_LOOKBACK = 252    # ~1 year of bars for the drawdown reference high
TREND_LOOKBACK = 20    # short-term trend window (~1 month)
WARN_DAYS = 30         # remember bearish/trim alerts this many calendar days

# Alert categories that are a caution for someone HOLDING the stock.
WARN_CATEGORIES = {"rsi_extended"}


def _round(v, dp=2):
    return None if v is None else round(float(v), dp)


def ticker_health(df, sector_state: str | None = None) -> dict | None:
    """Technical state from a single ticker's OHLCV. Pure/testable."""
    if df is None or df.empty or len(df) < 60:
        return None
    c = df["close"]
    px = float(c.iloc[-1])
    out: dict = {"close": _round(px)}

    s200 = sma(c, 200)
    if len(c) >= 200 and not s200.isna().iloc[-1]:
        v = float(s200.iloc[-1])
        out["sma200"] = _round(v)
        out["vs_sma200_pct"] = _round((px / v - 1) * 100)
    s50 = sma(c, 50)
    if len(c) >= 50 and not s50.isna().iloc[-1]:
        v = float(s50.iloc[-1])
        out["sma50"] = _round(v)
        out["vs_sma50_pct"] = _round((px / v - 1) * 100)

    r = rsi(c)
    if not r.isna().iloc[-1]:
        out["rsi"] = _round(float(r.iloc[-1]), 1)

    line, sig = macd(c)
    if not (line.isna().iloc[-1] or sig.isna().iloc[-1]):
        out["macd_bullish"] = bool(line.iloc[-1] > sig.iloc[-1])

    # trend: % change over the last TREND_LOOKBACK bars
    if len(c) > TREND_LOOKBACK:
        out["chg_20d_pct"] = _round((px / float(c.iloc[-1 - TREND_LOOKBACK]) - 1) * 100)

    # drawdown from the highest close in the lookback window
    window = c.tail(PEAK_LOOKBACK)
    peak = float(window.max())
    if peak > 0:
        out["peak_252d"] = _round(peak)
        out["drawdown_pct"] = _round((px / peak - 1) * 100)

    if sector_state:
        out["sector_state"] = sector_state
    return out


def recent_warnings(history: dict, bar_date: str, days: int = WARN_DAYS) -> dict[str, list[dict]]:
    """{ticker: [{date, category, rule, direction}]} for bearish crosses and
    trim alerts within the window — so a warning isn't lost the next day."""
    cutoff = (dt.date.fromisoformat(bar_date) - dt.timedelta(days=days)).isoformat()
    out: dict[str, list[dict]] = {}
    for day in history.get("days", []):
        for a in day.get("alerts", []):
            if a.get("date", "") < cutoff:
                continue
            if a.get("direction") == "bearish" or a.get("category") in WARN_CATEGORIES:
                rec = {"date": a["date"], "category": a.get("category"),
                       "rule": a.get("rule"), "direction": a.get("direction")}
                lst = out.setdefault(a["ticker"], [])
                if rec not in lst:
                    lst.append(rec)
    for lst in out.values():
        lst.sort(key=lambda r: r["date"], reverse=True)
    return out


def build(output_dir: Path = DEFAULT_OUTPUT_DIR, frames: dict | None = None,
          bar_date: str | None = None, sector_states: dict | None = None,
          history: dict | None = None, now: dt.datetime | None = None) -> dict:
    frames = frames or {}
    sector_states = sector_states or {}
    if history is None:
        try:
            history = json.loads((output_dir / "history.json").read_text())
        except (OSError, ValueError):
            history = {"days": []}

    tickers: dict[str, dict] = {}
    for sym, df in frames.items():
        h = ticker_health(df, sector_states.get(sym))
        if h is not None:
            tickers[sym] = h

    warnings = recent_warnings(history, bar_date) if bar_date else {}
    for sym, warns in warnings.items():
        if sym in tickers:
            tickers[sym]["recent_warnings"] = warns

    data = {
        "schema_version": SCHEMA_VERSION,
        "generated_at": (now or dt.datetime.now(dt.timezone.utc)).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "bar_date": bar_date,
        "warn_days": WARN_DAYS,
        "tickers": dict(sorted(tickers.items())),
    }
    output_dir.mkdir(parents=True, exist_ok=True)
    path = output_dir / "health.json"
    try:
        prev = json.loads(path.read_text())
    except (OSError, ValueError):
        prev = None
    if prev is not None and {**prev, "generated_at": None} == {**data, "generated_at": None}:
        data["generated_at"] = prev["generated_at"]
    path.write_text(json.dumps(data, sort_keys=True, indent=1) + "\n")
    return data
